src/test/crimson/perf_crimson_msgr.cc

```diff
@@ -266,46 +330,221 @@ static seastar::future<> run(
+               << std::setw(8) << (total_lat_s / sampled_count * 1000);
+          std::cout << "--------------"
+                    << " summary "
+                    << "--------------"
+                    << std::endl;
+          std::cout << sout.str() << std::endl << std::endl;
+        }
+      };
+
+      seastar::future<> report_period(TimerReport& report) {
+        return container().invoke_on_all([&report] (auto& client) {
+          if (client.is_active()) {
+            PeriodSession* snap = report.get_snap_by_job(client.sid);
+            client.period_session.reset(client.conn_session.received_count,
+                                        snap);
+            snap->depth = client.get_current_depth();
+          }
+        }).then([&report] {
+          report.report_period();
+        });
+      }
+
+      seastar::future<> report_summary(TimerReport& report) {
+        return container().invoke_on_all([&report] (auto& client) {
+          if (client.is_active()) {
+            ConnSession* summary = report.get_summary_by_job(client.sid);
+            *summary = client.conn_session;
+            summary->finish_time = mono_clock::now();
+          }
+        }).then([&report] {
+          report.report_summary();
+        });
+      }
+
+     public:
+      seastar::future<> dispatch_with_timer(unsigned ramp, unsigned seconds) {
+        logger().info("[all clients]: start sending MOSDOps from {} clients", jobs);
+        return container().invoke_on_all([] (auto& client) {
+          if (client.is_active()) {
+            client.do_dispatch_messages(client.active_conn.get());
+          }
+        }).then([this, ramp] {
+          logger().info("[all clients]: ramping up {} seconds...", ramp);
+          return seastar::sleep(std::chrono::seconds(ramp));
         }).then([this] {
-          logger().info("\nstart sending {} MOSDOps from {} clients",
-                        jobs * rounds, jobs);
-          mono_time start_time = mono_clock::now();
           return container().invoke_on_all([] (auto& client) {
             if (client.is_active()) {
-              return client.do_dispatch_messages(client.active_conn.get());
+              client.conn_session.start();
+              client.period_session.reset(client.conn_session.received_count);
             }
-            return seastar::now();
-          }).then([this, start_time] {
-            std::chrono::duration<double> dur_messaging = mono_clock::now() - start_time;
-            logger().info("\nSummary:\n  clients: {}\n  MOSDOps: {}\n  total time: {}s\n",
-                          jobs, jobs * rounds, dur_messaging.count());
+          });
+        }).then([this, seconds] {
+          logger().info("[all clients]: reporting {} seconds...\n", seconds);
+          return seastar::do_with(
+              TimerReport(jobs, seconds, msg_len), [this] (auto& report) {
+            report.report_header();
+            return seastar::do_until(
```

Comment:
could use `seastar::with_timeout()`?

Reply:
I think probably not, the `msgtime` is already controlled by `report.ticktock()` and the `elapsed` counter. It means that each report period is slightly more than 1s. So `seastar::with_timeout()` might cause unexpected interruptions.